core/services/synchronization/telemetry_ingress_batch_client.go

```diff
@@ -0,0 +1,301 @@
+
+	wgDone sync.WaitGroup
+	chDone chan struct{}
+
+	telemBufferSize   uint
+	telemMaxBatchSize uint
+	telemSendInterval time.Duration
+
+	workers      map[string]*telemetryIngressBatchWorker
+	workersMutex sync.Mutex
+}
+
+// NewTelemetryIngressBatchClient returns a client backed by wsrpc that
+// can send telemetry to the telemetry ingress server
+func NewTelemetryIngressBatchClient(url *url.URL, serverPubKeyHex string, ks keystore.CSA, logging bool, lggr logger.Logger, telemBufferSize uint, telemMaxBatchSize uint, telemSendInterval time.Duration) TelemetryIngressBatchClient {
+	return &telemetryIngressBatchClient{
+		telemBufferSize:   telemBufferSize,
+		telemMaxBatchSize: telemMaxBatchSize,
+		telemSendInterval: telemSendInterval,
+		url:               url,
+		ks:                ks,
+		serverPubKeyHex:   serverPubKeyHex,
+		globalLogger:      lggr,
+		logging:           logging,
+		lggr:              lggr.Named("TelemetryIngressBatchClient"),
+		chDone:            make(chan struct{}),
+		workers:           make(map[string]*telemetryIngressBatchWorker),
+	}
+}
+
+// Start connects the wsrpc client to the telemetry ingress server
+func (tc *telemetryIngressBatchClient) Start() error {
+	return tc.StartOnce("TelemetryIngressBatchClient", func() error {
+		privkey, err := tc.getCSAPrivateKey()
+		if err != nil {
+			return err
+		}
+
+		tc.connect(privkey)
+
+		return nil
+	})
+}
+
+// Close disconnects the wsrpc client from the ingress server and waits for all workers to exit
+func (tc *telemetryIngressBatchClient) Close() error {
+	return tc.StopOnce("TelemetryIngressBatchClient", func() error {
+		close(tc.chDone)
+		tc.wgDone.Wait()
+		return nil
+	})
+}
+
+func (tc *telemetryIngressBatchClient) connect(clientPrivKey []byte) {
+	tc.wgDone.Add(1)
+
+	go func() {
+		defer tc.wgDone.Done()
+
+		serverPubKey := keys.FromHex(tc.serverPubKeyHex)
+
+		conn, err := wsrpc.Dial(tc.url.String(), wsrpc.WithTransportCreds(clientPrivKey, serverPubKey))
+		if err != nil {
+			tc.lggr.Errorf("Error connecting to telemetry ingress server: %v", err)
+			return
+		}
+		defer conn.Close()
+
+		// Initialize a new wsrpc client caller
+		// This is used to call RPC methods on the server
+		if tc.telemClient == nil { // only preset for tests
+			tc.telemClient = telemPb.NewTelemClient(conn)
+		}
+
+		// Wait for close
+		<-tc.chDone
+
+	}()
+}
+
+// getCSAPrivateKey gets the client's CSA private key
+func (tc *telemetryIngressBatchClient) getCSAPrivateKey() (privkey []byte, err error) {
+	keys, err := tc.ks.GetAll()
+	if err != nil {
+		return privkey, err
+	}
+	if len(keys) < 1 {
+		return privkey, errors.New("CSA key does not exist")
+	}
+
+	return keys[0].Raw(), nil
+}
+
+// Send directs incoming telmetry messages to the worker responsible for pushing it to
+// the ingress server. If the worker telemetry buffer is full, messages are dropped
+// and a warning is logged.
+func (tc *telemetryIngressBatchClient) Send(payload TelemPayload) {
```

Comment:
This does not check if the client was connected or not (see `connect`).
At very least this should return error if the service is not connected.

Reply:
Let me know if my other comment https://github.com/smartcontractkit/chainlink/pull/5887#discussion_r799757259 doesn't address this, happy to discuss further